jerry-debugger/src/lib/debugger-client.ts

```diff
@@ -0,0 +1,404 @@
+      });
+
+      this.socket.on('error', (err) => {
+        reject(err);
+      });
+    });
+  }
+
+  disconnect() {
+    if (this.socket) {
+      this.socket.terminate();
+      this.socket = undefined;
+    }
+  }
+
+  abortConnection(message: string) {
+    this.disconnect();
+    console.log('Abort connection: ' + message);
+    throw new Error(message);
+  }
+
+  getFormatSize(format: string) {
+    let length = 0;
+    for (let i = 0; i < format.length; i++) {
+      switch (format[i]) {
+        case 'B':
+          length++;
+          break;
+
+        case 'C':
+          length += this.cpointerSize;
+          break;
+
+        case 'I':
+          length += 4;
+          break;
+
+        default:
+          throw new Error('unsupported message format');
+      }
+    }
+    return length;
+  }
+
+  decodeMessage(format: string, message: Uint8Array, offset = 0) {
+    // Format: B=byte I=int32 C=cpointer
+    // Returns an array of decoded numbers
+
+    const result = [];
+    let value;
+
+    if (offset + this.getFormatSize(format) > message.byteLength) {
+      this.abortConnection('received message too short');
+    }
+
+    for (let i = 0; i < format.length; i++) {
+      if (format[i] === 'B') {
+        result.push(message[offset]);
+        offset++;
+        continue;
+      }
+
+      if (format[i] === 'C' && this.cpointerSize === 2) {
+        if (this.littleEndian) {
+          value = message[offset] | (message[offset + 1] << 8);
+        } else {
+          value = (message[offset] << 8) | message[offset + 1];
+        }
+
+        result.push(value);
+        offset += 2;
+        continue;
+      }
+
+      assert(format[i] === 'I' || (format[i] === 'C' && this.cpointerSize === 4));
+
+      if (this.littleEndian) {
+        value = (message[offset] | (message[offset + 1] << 8)
+                 | (message[offset + 2] << 16) | (message[offset + 3] << 24));
+      } else {
+        value = ((message[offset] << 24) | (message[offset + 1] << 16)
+                 | (message[offset + 2] << 8) | message[offset + 3] << 24);
+      }
+
+      result.push(value);
+      offset += 4;
+    }
+
+    return result;
+  }
+
+  releaseFunction(message: Uint8Array) {
+  }
+
+  getBreakpoint(breakpointData: Array<number>) {
+    const func = this.functions[breakpointData[0]];
+    const offset = breakpointData[1];
+
+    if (offset in func.offsets) {
+      return {
+        breakpoint: func.offsets[offset],
+        at: true,
+      }
+    }
+
+    if (offset < func.firstBreakpointOffset) {
+      return {
+        breakpoint: func.offsets[func.firstBreakpointOffset],
+        at: true,
+      }
+    }
+
+    let nearestOffset = -1;
+    for (let currentOffset in func.offsets) {
+      const current = Number(currentOffset);
+      if ((current <= offset) && (current > nearestOffset)) {
+        nearestOffset = current;
+      }
+    }
+
+    return {
+      breakpoint: func.offsets[nearestOffset],
+      at: false,
+    };
```

Comment:
I think it would be nice to define `interface`s for each of the JrS debugger messages.

Reply:
Now this makes sense I think to have an interface defined for the 'options' to each delegate callback. I assume common sets of arguments could reuse the same interface: split them out later if something changes.